bot/bot.go

```diff
@@ -32,42 +32,95 @@ type Bot interface {
 	Channels() []string                                       // Channels returns list of channels which bot has joined.
 }
 
-// NewBot creates a new Bot implementation with a set of parameters.
-func NewBot(host string, ssl bool, nick, ident string, channels []string) Bot {
-	return NewBotWithProxy(host, ssl, nick, ident, channels, "")
-}
-
-// NewBotWithProxy creates a new Bot implementation with a set of parameters and a proxy.
-func NewBotWithProxy(host string, ssl bool, nick, ident string, channels []string, proxy string) Bot {
+// NewBotWithProxyPassword creates a new Bot implementation with required parameters and options.
+func NewBotOptions(host string, nick string, options ...func(*BotImpl)) (Bot, error) {
 	hostPort := strings.SplitN(host, ":", 2)
 	cfg := &client.Config{
-		Me:          &state.Nick{Nick: nick},
+		Me: &state.Nick{
+			Nick:  nick,
+			Ident: nick,
+			Name:  nick,
+		},
 		NewNick:     func(s string) string { return s + "_" },
 		PingFreq:    3 * time.Minute,
 		QuitMessage: "I have to go.",
 		Server:      host,
-		SSL:         ssl,
+		SSL:         false,
 		SSLConfig:   tls.Config(hostPort[0]),
 		Version:     "Powered by GoIRCBot",
 		Recover:     (*client.Conn).LogPanic,
 		SplitLen:    450,
-		Proxy:       proxy,
+		Proxy:       "",
+		Pass:        "",
 	}
-	cfg.Me.Ident = ident
-	cfg.Me.Name = nick
 
-	conn := client.Client(cfg)
-	conn.EnableStateTracking()
 	b := &BotImpl{
-		conn:      conn,
+		config:    cfg,
 		reconnect: true,
 		quit:      make(chan bool),
 		commands:  NewCommands(),
-		channels:  channels,
 	}
 
+	for _, option := range options {
+		option(b)
+	}
+
+	return b, nil
+}
+
+func SSL(ssl bool) func(*BotImpl) {
+	return func(b *BotImpl) { b.config.SSL = ssl }
+}
+
+func Ident(ident string) func(*BotImpl) {
+	return func(b *BotImpl) { b.config.Me.Ident = ident }
+}
+
+func RealName(realName string) func(*BotImpl) {
+	return func(b *BotImpl) { b.config.Me.Name = realName }
+}
+
+func Proxy(proxy string) func(*BotImpl) {
+	return func(b *BotImpl) { b.config.Proxy = proxy }
+}
+
+func Password(password string) func(*BotImpl) {
+	return func(b *BotImpl) { b.config.Pass = password }
+}
+
+func Channels(channels []string) func(*BotImpl) {
+	return func(b *BotImpl) { b.channels = channels }
+}
+
+// NewBot creates a new Bot implementation with a set of parameters.
```

Comment:
Can you mark `NewBot` and `NewBotWithProxy` deprecated?

`// DEPRECATED: use NewBotOptions instead.`

I want to clean this up at some point, potentially up to renaming `NewBotOptions` back to `NewBot`, maybe.